<a href="https://colab.research.google.com/github/fareed-asif/Student_Performance_Analysis/blob/main/SP24_BAI_031(PFAI_LAB_ASS_3).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("/content/StudentsPerformance.csv")

print("Shape (rows, columns):", df.shape)
print("\nFirst 5 rows:")
print(df.head())
print("\nColumn names:", list(df.columns))
print("\nData types:")
print(df.dtypes)

Shape (rows, columns): (1000, 8)

First 5 rows:
   gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  

Column names: ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 

In [2]:
score_cols = ['math score', 'reading score', 'writing score']

print("=== MEAN ===")
print(df[score_cols].mean().round(2))

print("\n=== MEDIAN ===")
print(df[score_cols].median())

print("\n=== MODE ===")
print(df[score_cols].mode().iloc[0])

print("\n=== STANDARD DEVIATION ===")
print(df[score_cols].std().round(2))

print("\n=== MIN ===")
print(df[score_cols].min())

print("\n=== MAX ===")
print(df[score_cols].max())

print("\n=== FULL SUMMARY ===")
print(df[score_cols].describe().round(2))

=== MEAN ===
math score       66.09
reading score    69.17
writing score    68.05
dtype: float64

=== MEDIAN ===
math score       66.0
reading score    70.0
writing score    69.0
dtype: float64

=== MODE ===
math score       65
reading score    72
writing score    74
Name: 0, dtype: int64

=== STANDARD DEVIATION ===
math score       15.16
reading score    14.60
writing score    15.20
dtype: float64

=== MIN ===
math score        0
reading score    17
writing score    10
dtype: int64

=== MAX ===
math score       100
reading score    100
writing score    100
dtype: int64

=== FULL SUMMARY ===
       math score  reading score  writing score
count     1000.00        1000.00        1000.00
mean        66.09          69.17          68.05
std         15.16          14.60          15.20
min          0.00          17.00          10.00
25%         57.00          59.00          57.75
50%         66.00          70.00          69.00
75%         77.00          79.00          79.00
max        100.00

In [3]:
print("=== Mean vs Median Comparison ===")
print(f"{'Subject':<20} {'Mean':>8} {'Median':>8} {'Difference':>12}")
print("-" * 50)

for col in score_cols:
    mean = df[col].mean()
    median = df[col].median()
    diff = mean - median
    print(f"{col:<20} {mean:>8.2f} {median:>8.2f} {diff:>12.2f}")

=== Mean vs Median Comparison ===
Subject                  Mean   Median   Difference
--------------------------------------------------
math score              66.09    66.00         0.09
reading score           69.17    70.00        -0.83
writing score           68.05    69.00        -0.95


In [4]:
print("=== Average Scores by Gender ===")
gender_stats = df.groupby('gender')[score_cols].mean().round(2)
print(gender_stats)

print("\n=== Count by Gender ===")
print(df['gender'].value_counts())

=== Average Scores by Gender ===
        math score  reading score  writing score
gender                                          
female       63.63          72.61          72.47
male         68.73          65.47          63.31

=== Count by Gender ===
gender
female    518
male      482
Name: count, dtype: int64


In [5]:
print("=== Average Scores by Parental Education ===")
parent_stats = df.groupby('parental level of education')[score_cols].mean().round(2)
print(parent_stats)

print("\n=== Count by Parental Education ===")
print(df['parental level of education'].value_counts())

=== Average Scores by Parental Education ===
                             math score  reading score  writing score
parental level of education                                          
associate's degree                67.88          70.93          69.90
bachelor's degree                 69.39          73.00          73.38
high school                       62.14          64.70          62.45
master's degree                   69.75          75.37          75.68
some college                      67.13          69.46          68.84
some high school                  63.50          66.94          64.89

=== Count by Parental Education ===
parental level of education
some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: count, dtype: int64


1. Female students score higher in reading (72.6) and writing (72.5) compared to males, while males score slightly higher in math (69.0 vs 63.6).

2. Students whose parents have a master's degree score the highest across all subjects (math: 69.7, reading: 75.4, writing: 75.7).

3. Mean and Median are very close for all three scores (difference < 1), indicating the data is approximately symmetric with no extreme skew.

4. Reading and writing scores are highly correlated — students who score high in one tend to score high in the other.

5. Standard deviation (~15 points) across all subjects shows significant variation in student performance.